src/envoy/utils/utils.cc

```diff
@@ -120,30 +143,37 @@ bool GetDestinationUID(const envoy::api::v2::core::Metadata& metadata,
 
 bool GetPrincipal(const Network::Connection* connection, bool peer,
                   std::string* principal) {
-  if (connection) {
-    Ssl::ConnectionInfo* ssl =
-        const_cast<Ssl::ConnectionInfo*>(connection->ssl());
-    if (ssl != nullptr) {
-      const std::vector<std::string> sans =
-          (peer ? ssl->uriSanPeerCertificate() : ssl->uriSanLocalCertificate());
-      if (sans.empty()) {
-        // empty result is not allowed
-        return false;
-      }
-      const std::string result = sans[0];
-      if (result.length() >= kSPIFFEPrefix.length() &&
-          result.compare(0, kSPIFFEPrefix.length(), kSPIFFEPrefix) == 0) {
-        // Strip out the prefix "spiffe://" in the identity.
-        *principal = result.substr(kSPIFFEPrefix.size());
-      } else {
-        *principal = result;
-      }
-      return true;
+  std::string result;
+  if (getRawPrincipal(connection, peer, &result)) {
+    if (hasSPIFFEPrefix(result)) {
+      // Strip out the prefix "spiffe://" in the identity.
+      *principal = result.substr(kSPIFFEPrefix.size());
```

Comment:
I think it's okay here as this function is used internally and it's the caller's responsibility to make sure `principal` is not null, otherwise they're making a mistake using this function.